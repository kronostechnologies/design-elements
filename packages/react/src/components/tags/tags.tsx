import React, { ReactElement, useMemo } from 'react';
import styled from 'styled-components';
import { IconName } from '../icon/icon';
import { TagSmall } from '../tag-small/tag-small';
import { TagMedium } from '../tag-medium/tag-medium';

const StyledUl = styled.ul`
    display: flex;
    list-style: none;
    margin: 0;
    padding: 0;
`;

type Size = 'small' | 'medium';

export interface Tag {
    iconName?: IconName;
    label: string;
    id?: string;
}

interface ListOption extends Tag {
    focusIndex: number;
    buttonRef: React.RefObject<HTMLButtonElement>
}

interface TagsProps {
    className?: string;
    options: Tag[];
    size?: Size;

    onDelete?:(option: Tag) => void;
}

function getListFromOptions(options: Tag[]): ListOption[] {
    return options.map(
        (option, index) => ({
            ...option,
            focusIndex: index,
            buttonRef: React.createRef<HTMLButtonElement>(),
        }),
    );
}

export function Tags({
    className, options, size = 'medium', onDelete,
}: TagsProps): ReactElement {
    const list: ListOption[] = useMemo(() => getListFromOptions(options), [options]);

    function handleDelete(listOption: ListOption): void {
        if (onDelete) {
            const previousOption = list[listOption.focusIndex - 1];

            onDelete(listOption);
            previousOption?.buttonRef.current?.focus();
        }
    }

    function getListItems(): JSX.Element[] {
        if (size === 'small') {
            return list.map(({ label }) => <TagSmall label={label} key={label} />);
        }

        return list.map((listOption) => (
            <TagMedium
                option={listOption}
                onDelete={onDelete ? handleDelete : undefined}
                key={listOption.id || listOption.label}
            />
        ));
    }

    return (
        <StyledUl className={className}>
            {getListItems()}
        </StyledUl>
    );
}
