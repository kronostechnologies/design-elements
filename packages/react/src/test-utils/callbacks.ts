export function doNothing(): void {
    // do nothing
}
