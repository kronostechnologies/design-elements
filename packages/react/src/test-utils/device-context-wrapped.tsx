import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
