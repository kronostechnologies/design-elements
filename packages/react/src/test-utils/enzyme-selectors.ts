import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
