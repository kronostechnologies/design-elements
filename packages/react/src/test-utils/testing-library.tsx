import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { DesignSystem, DesignSystemProps } from '../components/design-system';
import { equisoftTheme } from '../theme';

export function renderWithProviders(
    ui: Parameters<typeof render>[0],
    renderOptions: Omit<Parameters<typeof render>[1], 'wrapper'> | undefined = undefined,
    wrapperProps: DesignSystemProps = {},
): ReturnType<typeof render> {
    const {
        isolateStyles,
        language,
        staticDevice,
        theme,
    } = {
        isolateStyles: false,
        language: 'fr',
        theme: equisoftTheme,
        ...wrapperProps,
    };
    return render(
        ui,
        {
            ...renderOptions,
            wrapper: ({ children }) => (
                <MemoryRouter future={{ v7_startTransition: false, v7_relativeSplatPath: false }}>
                    <DesignSystem
                        isolateStyles={isolateStyles}
                        language={language}
                        staticDevice={staticDevice}
                        theme={theme}
                    >
                        {children}
                    </DesignSystem>
                </MemoryRouter>
            ),
        },
    );
}
