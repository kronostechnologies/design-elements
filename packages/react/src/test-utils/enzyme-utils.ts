import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
