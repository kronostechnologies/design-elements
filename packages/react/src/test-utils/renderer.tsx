import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
