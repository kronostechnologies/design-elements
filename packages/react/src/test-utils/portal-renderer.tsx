import { render, RenderResult } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { DeviceType } from '../components/device-context-provider/device-context-provider';
import { AllProviders } from './renderer';

export function renderPortalWithProviders(
    component: ReactElement,
    device?: DeviceType,
): RenderResult {
    return render(<AllProviders device={device}>{component}</AllProviders>);
}
