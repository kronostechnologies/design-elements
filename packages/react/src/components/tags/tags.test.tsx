import React from 'react';
import { getByTestId } from '../../test-utils/enzyme-selectors';
import { mountWithTheme, renderWithProviders } from '../../test-utils/renderer';
import { Tags, Tag } from './tags';

const tags: Tag[] = [
    { label: 'tag1' },
    { label: 'tag2' },
    { label: 'tag3' },
];

const tagsWithIcons: Tag[] = [
    { label: 'tag1', iconName: 'mail' },
    { label: 'tag2', iconName: 'home' },
    { label: 'tag3', iconName: 'calendar' },
];

describe('Tags', () => {
    test('previous item\'s delete-button should be focused when an item is deleted', () => {
        const wrapper = mountWithTheme(
            <Tags options={tags} onDelete={jest.fn()} />,
            { attachTo: document.body },
        );

        getByTestId(wrapper, 'tag3-delete-button').simulate('click');

        expect(document.activeElement).toBe(getByTestId(wrapper, 'tag2-delete-button').getDOMNode());
    });

    test('matches snapshot (size small)', () => {
        const tree = renderWithProviders(<Tags options={tags} size="small" />);

        expect(tree).toMatchSnapshot();
    });

    test('matches snapshot (size medium)', () => {
        const tree = renderWithProviders(<Tags options={tags} size="medium" />);

        expect(tree).toMatchSnapshot();
    });

    test('matches snapshot (with icons)', () => {
        const tree = renderWithProviders(<Tags options={tagsWithIcons} />);

        expect(tree).toMatchSnapshot();
    });

    test('matches snapshot (deletable)', () => {
        const tree = renderWithProviders(<Tags options={tags} onDelete={jest.fn()} />);

        expect(tree).toMatchSnapshot();
    });

    test('matches snapshot (mobile, small)', () => {
        const tree = renderWithProviders(<Tags options={tags} size='small' />, 'mobile');

        expect(tree).toMatchSnapshot();
    });

    test('matches snapshot (mobile, medium)', () => {
        const tree = renderWithProviders(<Tags options={tags} size="medium" />, 'mobile');

        expect(tree).toMatchSnapshot();
    });
});
