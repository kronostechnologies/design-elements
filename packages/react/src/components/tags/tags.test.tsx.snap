// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags matches snapshot (deletable) 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F1F2F2;
  border: 1px solid #878f9a;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  font-size: 0.75rem;
  line-height: 1.375rem;
  padding: 0 var(--spacing-1x);
}

.c1 + .c1 {
  margin-left: var(--spacing-1x);
}

.c3 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: var(--spacing-1x);
}

.c3:focus {
  outline: none;
}

.c3:focus {
  outline: none;
  box-shadow: 0 0 0 2px #00629666;
  box-shadow: 0 0 0 3px #00629666,0 0 0 1px #006296;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
}

<ul
  class="c0"
>
  <li
    class="c1 c2"
  >
    tag1
    <button
      aria-label="Remove tag1 from the list"
      class="c3"
      data-testid="tag1-delete-button"
      type="button"
    >
      <svg
        aria-hidden="true"
        color="#60666e"
        focusable="false"
        height="16"
        width="16"
      />
    </button>
  </li>
  <li
    class="c1 c2"
  >
    tag2
    <button
      aria-label="Remove tag2 from the list"
      class="c3"
      data-testid="tag2-delete-button"
      type="button"
    >
      <svg
        aria-hidden="true"
        color="#60666e"
        focusable="false"
        height="16"
        width="16"
      />
    </button>
  </li>
  <li
    class="c1 c2"
  >
    tag3
    <button
      aria-label="Remove tag3 from the list"
      class="c3"
      data-testid="tag3-delete-button"
      type="button"
    >
      <svg
        aria-hidden="true"
        color="#60666e"
        focusable="false"
        height="16"
        width="16"
      />
    </button>
  </li>
</ul>
`;

exports[`Tags matches snapshot (mobile, medium) 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F1F2F2;
  border: 1px solid #878f9a;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  font-size: 0.875rem;
  line-height: 1.875rem;
  padding: 0 var(--spacing-1x);
}

.c1 + .c1 {
  margin-left: var(--spacing-1x);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
}

<ul
  class="c0"
>
  <li
    class="c1 c2"
  >
    tag1
  </li>
  <li
    class="c1 c2"
  >
    tag2
  </li>
  <li
    class="c1 c2"
  >
    tag3
  </li>
</ul>
`;

exports[`Tags matches snapshot (mobile, small) 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F1F2F2;
  border: 1px solid #878f9a;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  font-size: 0.875rem;
  line-height: 1.375rem;
  padding: 0 var(--spacing-1x);
}

.c1 + .c1 {
  margin-left: var(--spacing-1x);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
}

<ul
  class="c0"
>
  <li
    class="c1 c2"
  >
    tag1
  </li>
  <li
    class="c1 c2"
  >
    tag2
  </li>
  <li
    class="c1 c2"
  >
    tag3
  </li>
</ul>
`;

exports[`Tags matches snapshot (size medium) 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F1F2F2;
  border: 1px solid #878f9a;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  font-size: 0.75rem;
  line-height: 1.375rem;
  padding: 0 var(--spacing-1x);
}

.c1 + .c1 {
  margin-left: var(--spacing-1x);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
}

<ul
  class="c0"
>
  <li
    class="c1 c2"
  >
    tag1
  </li>
  <li
    class="c1 c2"
  >
    tag2
  </li>
  <li
    class="c1 c2"
  >
    tag3
  </li>
</ul>
`;

exports[`Tags matches snapshot (size small) 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F1F2F2;
  border: 1px solid #878f9a;
  border-radius: 2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  font-size: 0.75rem;
  line-height: 0.875rem;
  padding: 0 var(--spacing-half);
}

.c1 + .c1 {
  margin-left: var(--spacing-1x);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
}

<ul
  class="c0"
>
  <li
    class="c1 c2"
  >
    tag1
  </li>
  <li
    class="c1 c2"
  >
    tag2
  </li>
  <li
    class="c1 c2"
  >
    tag3
  </li>
</ul>
`;

exports[`Tags matches snapshot (with icons) 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F1F2F2;
  border: 1px solid #878f9a;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  font-size: 0.75rem;
  line-height: 1.375rem;
  padding: 0 var(--spacing-1x);
}

.c1 + .c1 {
  margin-left: var(--spacing-1x);
}

.c3 {
  color: #60666e;
  margin-right: var(--spacing-1x);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
}

<ul
  class="c0"
>
  <li
    class="c1 c2"
  >
    <svg
      aria-hidden="false"
      aria-label="mail"
      class="c3"
      color="currentColor"
      data-testid="tag1-icon"
      focusable="false"
      height="16"
      width="16"
    />
    tag1
  </li>
  <li
    class="c1 c2"
  >
    <svg
      aria-hidden="false"
      aria-label="home"
      class="c3"
      color="currentColor"
      data-testid="tag2-icon"
      focusable="false"
      height="16"
      width="16"
    />
    tag2
  </li>
  <li
    class="c1 c2"
  >
    <svg
      aria-hidden="false"
      aria-label="calendar"
      class="c3"
      color="currentColor"
      data-testid="tag3-icon"
      focusable="false"
      height="16"
      width="16"
    />
    tag3
  </li>
</ul>
`;
